README.md
# hermes

Authenticate on linux by plugging your USB stick!

Demo: https://www.youtube.com/watch?v=W7L9FrIDYgU

## Installation

### Installing hermes

*Note: for now, only a debian/ubuntu package is provided.*

Download the [debian package][0] and install it:

```
$ sudo dpkg -i hermes_1.8_amd64.deb
$ sudo apt-get -f install
```

### Preparing the usb stick

I recommend having 2 partitions on your usb stick: a very small one
(hermes needs 266 bytes as of now, so allocate 2048 bytes to be safe),
and another partition for your data.

Let's assume that the partition for hermes is `/dev/sdb1`, and your
user is `foo`.

To make your usb stick an hermes device, run:

```
$ sudo hermes write /dev/sdb1 foo
```

You can see what `hermes` can do with `hermes help`. (Although for
now, it can only run the write command seen above.)

### Done!

You can now do exactly as in the video. When you need to login, plug
in the usb stick, hit enter with an empty password, and you should be
logged in!

## Manual installation

If you're not on debian, here are the steps to manually install
hermes.

Download the following (equivalent to your distribution) dependencies:
gcc, sbcl, wget, buildapp, libpam0g-dev, libfixposix-dev.

Then run:

```
# Build hermes
$ git clone https://github.com/ralt/hermes
$ cd hermes/
$ make

# Install the built files
$ sudo make install

# Configure your system
$ sudo pam-auth-update # Select "Hermes"
$ sudo addgroup hermes
```

If you're using systemd, you can then run the following:

```
$ sudo systemctl enable hermes
$ sudo systemctl start hermes
```

If you're not using systemd, please look at the
`debian/services/hermes.service` file and reproduce the equivalent for
your init system.

From then on, you can follow the steps in
[Preparing the usb stick][1].

## How does it work?

A string of 128 characters is put on the both usb stick and in the
user's home directory. When connection time comes, the tokens are
checked to see if they match. If they do, the user can be connected.

Every time there is a successful login, the token is regenerated, for
security reasons.

## Developer todo

- better error handling. Right now it's fprintf(stderr) or perror()
  everywhere, randomly. Make it consistent.

## Technical process

The `hermes write <device> <user>` command will generate a 128-bytes
token, and write it to both `/etc/hermes/<user>` and the device. On
the device, it will first write 5 bytes to recognize an hermes device.

When login is needed, the PAM module will send the user to the
/var/run/hermes.sock UNIX socket. The hermes-daemon service will then
look in /dev/sd* for a hermes device (starts with bytes 82, 111, 98,
105, 110). If none exists, it sends "false" to the UNIX socket. If one
exists, it compares with the token in `/etc/hermes/<user>`, and sends
either "true" or "false" back to the UNIX socket. The PAM module just
has to get the boolean result back and it can return it.

## License

The project is provided under the MIT license.


  [0]: https://github.com/ralt/hermes/releases/download/1.8/hermes_1.8_amd64.deb
  [1]: https://github.com/ralt/hermes#preparing-the-usb-stick


debian/changelog
hermes (1.8) stable; urgency=medium

  * Really fix make install

 -- Florian Margaine <florian@margaine.com>  Sun, 27 Dec 2015 23:08:36 +0200

hermes (1.7) stable; urgency=medium

  * Fix make install

 -- Florian Margaine <florian@margaine.com>  Sun, 27 Dec 2015 23:03:35 +0200

hermes (1.6) stable; urgency=medium

  * Retries more often with less delay

 -- Florian Margaine <florian@margaine.com>  Sun, 18 Oct 2015 19:06:45 +0200

hermes (1.5) stable; urgency=medium

  * Adds overrideable variables
  * Internal refactoring

 -- Florian Margaine <florian@margaine.com>  Sun, 04 Oct 2015 16:48:49 +0200

hermes (1.4) stable; urgency=medium

  * Tries several times over 5 seconds to find an hermes device

 -- Florian Margaine <florian@margaine.com>  Mon, 28 Sep 2015 09:25:50 +0200

hermes (1.3) stable; urgency=medium

  * Adds safe one-time tokens.
  * Adds syslog facilities
  * Adds user to group on hermes write

 -- Florian Margaine <florian@margaine.com>  Thu, 24 Sep 2015 16:17:55 +0200

hermes (1.2) unstable; urgency=medium

  * Adds one-time tokens.
  * Moves to Common-Lisp based daemon

 -- Florian Margaine <florian@margaine.com>  Mon, 21 Sep 2015 01:04:22 +0200

hermes (1.1) stable; urgency=medium

  * Moves to global tokens, allowing pre-decryption tokens.

 -- Florian Margaine <florian@margaine.com>  Sun, 20 Sep 2015 20:21:55 +0200

hermes (1.0) stable; urgency=low

  * Initial release

 -- Florian Margaine <florian@margaine.com>  Fri, 04 Sep 2015 15:29:45 +0200


hermes-service.c
/*
  The service, run as root, will listen on the /var/run/hermes.sock
  socket. It is waiting for a user char* to come in, after which it
  will check if the user can login. It then sends a boolean back into
  the UNIX socket.
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdint.h>
#include <unistd.h>
#include <sys/stat.h>
#include <grp.h>
#include <errno.h>
#include <glob.h>
#include <arpa/inet.h>
#include <syscall.h>
#include <linux/random.h>

#define FINGERPRINT_LENGTH 5
#define TOKEN_LENGTH 128
#define MAX_USERNAME_LENGTH 33

static int globerr(const char*, int);
static bool is_block_device(const char*);
static bool has_hermes_fingerprint(const char*);
static bool is_hermes_device(const char*);
static bool can_login(char[MAX_USERNAME_LENGTH]);
static bool regenerate_token(char user[MAX_USERNAME_LENGTH]);
static bool find_hermes_device(char**);
static bool read_device_token(char*, uint8_t[TOKEN_LENGTH]);
static bool read_user_token(char user[MAX_USERNAME_LENGTH],
			    uint8_t token[TOKEN_LENGTH]);
static bool timing_safe_compare(uint8_t[TOKEN_LENGTH], uint8_t[TOKEN_LENGTH]);
static bool get_random_token(uint8_t token[TOKEN_LENGTH]);
static bool write_device_token(char *hermes_device, uint8_t token[TOKEN_LENGTH]);
static bool write_user_token(char *user, uint8_t token[TOKEN_LENGTH]);

int main(int argc, char *argv[])
{
	const char *socket_path = "/var/run/hermes.sock";
	struct sockaddr_un addr;

	/* usernames are max 32 chars + 1 for null char */
	char user[MAX_USERNAME_LENGTH];

	int fd, cl, rc;

	if ((fd = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
		perror("socket");
		exit(EXIT_FAILURE);
	}

	memset(&addr, 0, sizeof(addr));
	addr.sun_family = AF_UNIX;
	strncpy(addr.sun_path, socket_path, sizeof(addr.sun_path) - 1);

	unlink(socket_path);

	if (bind(fd, (const struct sockaddr*) &addr, sizeof(addr)) == -1)
	{
		perror("bind");
		exit(EXIT_FAILURE);
	}

	if (chmod(socket_path, S_IRGRP | S_IWGRP) == -1)
	{
		fprintf(stderr, "%s: can't chmod %s\n", strerror(errno), socket_path);
		exit(EXIT_FAILURE);
	}

	struct group *hermes_group = getgrnam("hermes");
	if (hermes_group == NULL)
	{
		perror("hermes group not found");
		exit(EXIT_FAILURE);
	}

	if (chown(socket_path, 0, hermes_group->gr_gid) == -1)
	{
		fprintf(stderr, "%s: can't chown %s to hermes group\n",
			strerror(errno), socket_path);
		exit(EXIT_FAILURE);
	}

	if (listen(fd, 5) == -1)
	{
		perror("listen");
		exit(EXIT_FAILURE);
	}

	while (true)
	{
		if ((cl = accept(fd, NULL, NULL)) == -1)
		{
			perror("accept");
			continue;
		}

		while ((rc = read(cl, &user, MAX_USERNAME_LENGTH)) > 0)
		{
			bool result = can_login(user);

			if (result)
			{
				result = regenerate_token(user);
			}

			if (write(cl, &result, sizeof(bool)) != 1)
			{
				perror("write result");
				exit(EXIT_FAILURE);
			}
		}

		if (rc == -1)
		{
			perror("read");
			exit(EXIT_FAILURE);
		}

		if (rc == 0)
		{
			continue;
		}
	}

	if (close(fd) != 0)
	{
		exit(EXIT_FAILURE);
	}

	return EXIT_SUCCESS;
}

static bool can_login(char user[MAX_USERNAME_LENGTH])
{
	bool retval = false;

	char *hermes_device;
	bool device_found = find_hermes_device(&hermes_device);
	if (!device_found)
	{
		return false;
	}

	uint8_t device_token[TOKEN_LENGTH];
	if (!read_device_token(hermes_device, device_token))
	{
		goto clean_exit;
	}

	uint8_t user_token[TOKEN_LENGTH];
	if (!read_user_token(user, user_token))
	{
		goto clean_exit;
	}

	retval = timing_safe_compare(device_token, user_token);

clean_exit:
	free(hermes_device);

	return retval;
}

static bool regenerate_token(char user[MAX_USERNAME_LENGTH])
{
	bool retval = false;

	uint8_t new_token[TOKEN_LENGTH];
	if (!get_random_token(new_token))
	{
		return false;
	}

	char *hermes_device;
	bool device_found = find_hermes_device(&hermes_device);
	if (!device_found)
	{
		return false;
	}

	if (!write_device_token(hermes_device, new_token))
	{
		goto clean_exit;
	}

	if (!write_user_token(user, new_token))
	{
		/* @TODO
		   There should be a way to recover here.
		   Because at this point, the new token is written on
		   the device, but not in the user's file. A very
		   inconsistent state. */
		goto clean_exit;
	}

	retval = true;

clean_exit:
	free(hermes_device);

	return retval;
}

static bool get_random_token(uint8_t token[TOKEN_LENGTH])
{
	if (syscall(SYS_getrandom, token, TOKEN_LENGTH, GRND_NONBLOCK) !=
	    TOKEN_LENGTH)
	{
		perror("getrandom");
		return false;
	}
	return true;
}

static bool write_device_token(char *path, uint8_t token[TOKEN_LENGTH])
{
	bool retval = false;

	FILE *fd;
	size_t written_bytes;

	fd = fopen(path, "wb");
	if (fd == NULL)
	{
		fprintf(stderr, "%s: can't write to %s\n", strerror(errno), path);
		return false;
	}

	if ((fseek(fd, FINGERPRINT_LENGTH, 0)) != 0)
	{
		fprintf(stderr, "%s: can't fseek into %s\n", strerror(errno), path);
		goto clean_exit;
	}

	written_bytes = fwrite(token,
			       sizeof(uint8_t),
			       TOKEN_LENGTH,
			       fd);
	if (written_bytes != TOKEN_LENGTH)
	{
		fprintf(stderr, "%s: couldn't write to %s\n", strerror(errno), path);
		goto clean_exit;
	}

	retval = true;

clean_exit:
	if (fclose(fd) != 0)
	{
		fprintf(stderr, "%s: couldn't close %s\n", strerror(errno), path);
		retval = false;
	}

	return retval;
}

static bool write_user_token(char *user, uint8_t token[TOKEN_LENGTH])
{
	const char *prefix = "/etc/hermes/";
	char user_token_path[strlen(prefix) + MAX_USERNAME_LENGTH];

	memcpy(user_token_path, prefix, strlen(prefix));
	memcpy(user_token_path + strlen(prefix), user, MAX_USERNAME_LENGTH);

	bool retval = false;
	FILE *fd;
	size_t written_bytes;

	fd = fopen(user_token_path, "wb");
	if (fd == NULL)
	{
		fprintf(stderr, "%s: can't write to %s\n", strerror(errno),
			user_token_path);
		return false;
	}

	written_bytes = fwrite(token,
			       sizeof(uint8_t),
			       TOKEN_LENGTH,
			       fd);
	if (written_bytes != TOKEN_LENGTH)
	{
		fprintf(stderr, "%s: couldn't write to %s\n", strerror(errno),
			user_token_path);
		goto clean_exit;
	}

	retval = true;

clean_exit:
	if (fclose(fd) != 0)
	{
		fprintf(stderr, "%s: couldn't close %s\n", strerror(errno),
			user_token_path);
		retval = false;
	}

	return retval;
}

static bool find_hermes_device(char **hermes_device)
{
	glob_t files;
	bool device_found = false;

	if (glob("/dev/sd*", GLOB_ERR | GLOB_NOSORT, globerr, &files) != 0)
	{
		return false;
	}

	for (size_t i = 0; i < files.gl_pathc; i++)
	{
		if (is_hermes_device(files.gl_pathv[i]))
		{
			*hermes_device = malloc(strlen(files.gl_pathv[i]) *
					       sizeof(uint8_t));
			if (*hermes_device == NULL)
			{
				perror("malloc hermes_device");
				exit(EXIT_FAILURE);
			}

			memcpy(*hermes_device,
			       files.gl_pathv[i],
			       strlen(files.gl_pathv[i]) + 1);

			device_found = true;
			break;
		}
	}

	globfree(&files);

	return device_found;
}

static bool read_device_token(char *path, uint8_t token[TOKEN_LENGTH])
{
	FILE *fd;
	size_t bytes_read;
	bool retval = false;

	fd = fopen(path, "rb");
	if (fd == NULL)
	{
		fprintf(stderr, "%s: can't read %s\n", strerror(errno), path);
		return false;
	}

	if ((fseek(fd, FINGERPRINT_LENGTH, 0)) != 0)
	{
		fprintf(stderr, "%s: can't fseek %d bytes in %s\n",
			strerror(errno), FINGERPRINT_LENGTH, path);
		goto safe_close;
	}

	bytes_read = fread(token,
			   sizeof(uint8_t),
			   TOKEN_LENGTH,
			   fd);
	if (bytes_read != TOKEN_LENGTH)
	{
		fprintf(stderr, "%s: can't read the token\n", strerror(errno));
		goto safe_close;
	}

	retval = true;

 safe_close:
	if (fclose(fd) != 0)
	{
		fprintf(stderr, "%s: can't close %s\n", strerror(errno), path);
	}

	return retval;
}

static bool read_user_token(char user[MAX_USERNAME_LENGTH],
			    uint8_t token[TOKEN_LENGTH])
{
	const char *prefix = "/etc/hermes/";
	char user_token_path[strlen(prefix) + MAX_USERNAME_LENGTH];

	memcpy(user_token_path, prefix, strlen(prefix));
	memcpy(user_token_path + strlen(prefix), user, MAX_USERNAME_LENGTH);

	FILE *fd;
	size_t bytes_read;
	bool retval = false;

	fd = fopen(user_token_path, "rb");
	if (fd == NULL)
	{
		perror("open user_token_path");
		return false;
	}

	bytes_read = fread(token, sizeof(uint8_t), TOKEN_LENGTH, fd);
	if (bytes_read != TOKEN_LENGTH)
	{
		perror("read token");
		goto safe_close;
	}

	retval = true;

 safe_close:
	if (fclose(fd) != 0)
	{
		perror("close user_token_path");
	}

	return retval;
}

static int globerr(const char *path, int eerrno)
{
	return true;
}

static bool is_block_device(const char *path)
{
	int retval;
	struct stat sb;

	retval = stat(path, &sb);
	if (retval == -1)
	{
		return false;
	}

	return S_ISBLK(sb.st_mode);
}

static bool has_hermes_fingerprint(const char *path)
{
	bool ret;
	FILE *fd;
	size_t bytes_read;
	uint8_t bytes[FINGERPRINT_LENGTH];
	const uint8_t expected_bytes[FINGERPRINT_LENGTH] = { 82, 111, 98, 105, 110 };

	fd = fopen(path, "rb");
	if (fd == NULL)
	{
		fprintf(stderr, "%s: can't read %s\n", strerror(errno), path);
		return false;
	}

	bytes_read = fread(&bytes, sizeof(uint8_t), FINGERPRINT_LENGTH, fd);
	if (bytes_read < 1)
	{
		fprintf(stderr, "%s: can't read enough bytes\n", strerror(errno));
		ret = false;
		goto safe_exit;
	}

	for (size_t i = 0; i < bytes_read; i++)
	{
		if (bytes[i] != expected_bytes[i])
		{
			ret = false;
			goto safe_exit;
		}
	}

	ret = true;
	goto safe_exit;

safe_exit:
	if (fclose(fd) != 0)
	{
		fprintf(stderr, "%s: can't close %s\n", strerror(errno), path);
		return false;
	}

	return ret;
}

static bool is_hermes_device(const char *path)
{
	if (!is_block_device(path))
	{
		return false;
	}

	if (!has_hermes_fingerprint(path))
	{
		return false;
	}

	return true;
}

static bool timing_safe_compare(uint8_t device_token[TOKEN_LENGTH],
				uint8_t user_token[TOKEN_LENGTH])
{
	int result = 0;

	for (size_t i = 0; i < TOKEN_LENGTH; i++) {
		result |= device_token[i % TOKEN_LENGTH] ^ user_token[i];
	}

	return result == 0;
}
